### PR TITLE
feat(v2): 🔥 🔥 better loading UX ⚡️

### PR DESCRIPTION
**Pull request by endiliey**
## TL;DR: see test plan or check https://deploy-preview-1383--docusaurus-2.netlify.com/

When routes take a long time to load, we’ll want to display some sort of loading indicator to the user. 
Example:
suspense-loading 6edec3cf

**However, if our internet speed is super slow. It can be very annoying to see a blank page and a spinner** 

What we'd really like to do is to display a loading bar over the current page while the next route loads… well, unless the transition only takes 100ms. Then you probably just want to keep displaying the current page until the next one is ready, because showing a loading bar for only 100ms also looks bad
no-delay-loading c22855b3

Inspired by YouTube and Medium style progress bar, here is what the flow for our app looks like

1. When routes navigation happens (e.g: when you click `<Link />` component, we trick the user by rendering the old route screen while trying to preload the next route screen. Only after the promise is resolved, we renders the next route screen.
2. However, if within one second it still doesnt resolve, a loading progress bar (like medium/youtube) is displayed to make user perceive there is some loading state.
3. The 1 second timeout is based on https://www.nngroup.com/articles/response-times-3-important-limits/. In fact, I copied the timeout from Gatsby 😄 .

You can try opening this PR's netlify, it is almost impossible now to see blank page and a spinner. 
See test plan for comparison

Others:
- Remove react-perimeter and react-listener (just use native mouse hover)

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/Docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Throttle internet speed to slow.

**Before**
loading screen -before

**After**
Better loading UX. Youtube-style progress bar while still rendering old screen.
loading screen -after

If the page transition is <1s, there is no progress bar being shown. Making the site 🔥 ⚡️ blazingly fast
amazingly fast

### CODE REVIEW

File: packages/docusaurus/lib/client/exports/Link.js [hunk 1]
```diff
@@ -6,15 +6,15 @@
...
 import {NavLink} from 'react-router-dom';
 
 const internalRegex = /^\/(?!\/)/;
 
 function Link(props) {
-  const {to, href, preloadProximity = 20} = props;
+  const {to, href} = props;
   const targetLink = to || href;
   const isInternal = internalRegex.test(targetLink);
+  let preloaded = false;
```
Comment (yangshun) [inline packages/docusaurus/lib/client/exports/Link.js]: This means that the `preloaded` state doesn't persist across renders. Is that intended?

File: packages/docusaurus/lib/client/pendingNavigation.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import React from 'react';
+import {Route, withRouter} from 'react-router-dom';
+import nprogress from 'nprogress';
+import 'nprogress/nprogress.css';
+import preload from './preload';
+
+let progressBarTimeout = null;
```
Comment (yangshun) [inline packages/docusaurus/lib/client/pendingNavigation.js]: This should be part of the component. If we put it here it means all instances of the component will share the same timeout. Even though there should only be one global spinner, it's still a better practice to put the timeout in the component.

```js
class ProgressBar extends React.Component {
  ...
  progressBarTimeout = null
}
```

File: packages/docusaurus/lib/client/pendingNavigation.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+  componentWillReceiveProps(nextProps) {
+    const navigated = nextProps.location !== this.props.location;
+    const {routes} = this.props;
+
+    if (navigated) {
+      window.scrollTo(0, 0);
+
+      // Only show the loading bar if it has passed 1 second
+      // see https://www.nngroup.com/articles/response-times-3-important-limits/
+      startProgressBar(1000);
```
Comment (yangshun) [inline packages/docusaurus/lib/client/pendingNavigation.js]: Make this a prop and default value of 1000?

File: packages/docusaurus/lib/client/pendingNavigation.js [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+      // load data while the old screen remains
+      preload(routes, nextProps.location.pathname)
+        .then(() => {
+          this.setState({
+            previousLocation: null,
+          });
+          stopProgressBar();
+        })
+        .catch(e => console.log(e));
```
Comment (yangshun) [inline packages/docusaurus/lib/client/pendingNavigation.js]: `console.warn`

File: packages/docusaurus/lib/client/pendingNavigation.js [hunk 5]
```diff
@@ -0,0 +1,83 @@
+/**
```
Comment (yangshun) [inline packages/docusaurus/lib/client/pendingNavigation.js]: This file name should be `PendingNavigation` as it's a component.

File: packages/docusaurus/lib/default-theme/Loading/styles.module.css [hunk 6]
```diff
@@ -1,9 +1,9 @@
-/**
- * Copyright (c) 2017-present, Facebook, Inc.
- *
- * This source code is licensed under the MIT license found in the
- * LICENSE file in the root directory of this source tree.
- */
+/**
```
Comment (yangshun) [inline packages/docusaurus/lib/default-theme/Loading/styles.module.css]: Seems like random spaces added here. Did you configure your editor to strip trailing spaces?

File: packages/docusaurus/lib/client/pendingNavigation.js [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        previousLocation: this.props.location,
+      });
+
+      // load data while the old screen remains
+      preload(routes, nextProps.location.pathname)
+        .then(() => {
+          this.setState({
+            previousLocation: null,
+          });
+          stopProgressBar();
```
Comment (yangshun) [inline packages/docusaurus/lib/client/pendingNavigation.js]: Use callback style

```
this.setState({
  previousLocation: null,
}, stopProgressBar);
```

File: packages/docusaurus/lib/client/exports/Link.js (hunk 1 above)
Comment (endiliey) [reply packages/docusaurus/lib/client/exports/Link.js]: yes. It's okay anyway since the dedupe is actually happening on docusaurus.js
